src/com/amazon/corretto/crypto/provider/RsaGen.java

```diff
@@ -66,11 +67,17 @@ public void initialize(int keysize, SecureRandom rnd) throws InvalidParameterExc
     }
   }
 
-  private static RSAKeyGenParameterSpec validateParameter(RSAKeyGenParameterSpec spec)
+  private static RSAKeyGenParameterSpec validateParameter(final RSAKeyGenParameterSpec spec)
       throws InvalidAlgorithmParameterException {
-    if (spec.getKeysize() < 512) {
+    if (Loader.FIPS_BUILD && spec.getPublicExponent().compareTo(RSAKeyGenParameterSpec.F4) <= -1) {
```

Comment:
Earlier, you say this:
> RSA_generate_key_fips doesn't allow  ... public exponents other than F4

That suggests that we should be checking for equality, not a minimum exponent size. Which is correct?

Reply:
AWS-LC doesn't follow BC-FIPS behavior in this case. Here I'm following BC-FIPS behavior.

Reply:
i think L72 is the public exp equality check and L77 is the key size check. from [BigInteger docs](https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html#compareTo(java.math.BigInteger)):

> Compares this BigInteger with the specified BigInteger. This method is provided in preference to individual methods for each of the six boolean comparison operators (<, ==, >, >=, !=, <=). The suggested idiom for performing these comparisons is: (x.compareTo(y) <op> 0), where <op> is one of the six comparison operators.

by that convention, perhaps this should be

```
spec.getPublicExponent().compareTo(RSAKeyGenParameterSpec.F4) != 0
```

or just use `.equals` (we risk NPE if proper input validation isn't done on `spec`)

```
!RSAKeyGenParameterSpec.F4.equals(spec.getPublicExponent())
```

?

Reply:
hm, so FIPS allows non-standard public exponents as long as they're >= 65537?

Reply:
I wouldn't say FIPS, but BC-FIPS yes.

Reply:
@geedo0 I added comments to clarify the behavior.